### PR TITLE
[DD4hep] MagneticField Builder Shapes Use Cleanup - Use dd4hep::Solid Directly

### PR DESCRIPTION
**Pull request by ianna**
#### PR description:

Addressing https://github.com/cms-sw/cmssw/issues/31569

* Use DD4hep Shapes and their accessors

This is done in preparation of removing the extra classes that duplicate the DD4hep ones:
https://cmssdt.cern.ch/lxr/source/DetectorDescription/DDCMS/interface/DDShapes.h

@namapane - The functionality is unchanged - essentially, it's a one to one copy from DDCMS. The comparison tests pass.
The `PseudoTrap` workaround is untouched. Please, check the units conversion and the angle restriction. 

If it's ok with you, I'd like to integrate this PR asap. It will allow me to proceed with removal of the `DDShapes.h` and `DDShapes.cc`. The MF builder is the last one that using it.

There is an inconsistency in getting a half Z (https://github.com/AIDASoft/DD4hep/issues/714). We may need to revisit this code depending on the issue outcome later.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by ianna** [comment]
please test

**Comment by namapane** [comment]
Uhm. buildBox.icc etc. are also used in the non-DD4Hep version that is still the default. I am a bit uncomfortable with diverging in the two implementations, unless there is a clear timeline on when we will make the DD4Hep version default and drop the non-DD4Hep one.
Also I would have to test all existing geometries built from both DB and xml to check for possible side effects, can you wait for a few days for these checks before merging ?

**Comment by namapane** [comment]
And actually what is the exact point of moving the content of the .icc files in the cc?  From a quick look I did not spot code changes... I would avoid duplicating this code if that is not strictly necessary.

**Comment by ianna** [comment]
> Uhm. buildBox.icc etc. are also used in the non-DD4Hep version that is still the default. I am a bit uncomfortable with diverging in the two implementations, unless there is a clear timeline on when we will make the DD4Hep version default and drop the non-DD4Hep one.
> Also I would have to test all existing geometries built from both DB and xml to check for possible side effects, can you wait for a few days for these checks before merging ?

@namapane - Sure, no problem. As far as I understand, the 20th October is a deadline to start full validation. I'd like to restrict the API beforehand.
 
The differences are the shape APIs. For example, here is a new version:
```c++
  DDBox box(solid.solid());
  // Old DD needs mm to cm conversion, but DD4hep needs no conversion.
  // convertUnits should be defined appropriately.
  double halfX = convertUnits(box.x());
  double halfY = convertUnits(box.y());
  double halfZ = convertUnits(box.z());
```
vs an old version:
```c++
  DDBox box(solid);
  // Old DD needs mm to cm conversion, but DD4hep needs no conversion.
  // convertUnits should be defined appropriately.
  double halfX = convertUnits(box.halfX());
  double halfY = convertUnits(box.halfY());
  double halfZ = convertUnits(box.halfZ());
```

After https://github.com/cms-sw/cmssw/pull/31577 and https://github.com/cms-sw/cmssw/pull/31571 are merged MF builder remains the only package that is still using the DDShapes wrapper classes.

If you prefer, we can move it under your control :-) https://cmssdt.cern.ch/lxr/source/DetectorDescription/DDCMS/interface/DDShapes.h
https://cmssdt.cern.ch/lxr/source/DetectorDescription/DDCMS/src/DDShapes.cc

In any case, you are welcome to take over this PR.

**Comment by namapane** [comment]
@ianna I see; I would actually prefer to modify the existing functions to pass the list of parameters for each case (the solution that was eventually adopted for buildPseudoTrap) so that we don't need to duplicate code. It should be rather straightforward but I need to carve out a couple of days for that... What is your target date for restricting the API?

What is your deadline to restrict the API?

**Comment by ianna** [comment]
> @ianna I see; I would actually prefer to modify the existing functions to pass the list of parameters for each case (the solution that was eventually adopted for buildPseudoTrap) so that we don't need to duplicate code. It should be rather straightforward but I need to carve out a couple of days for that... What is your target date for restricting the API?
> 
> What is your deadline to restrict the API?

@namapane - I was hoping to finish it this week since I need to concentrate on another unrelated project, but it can wait a couple of days, for sure.

File: MagneticField/GeomBuilder/src/DD4hep_volumeHandle.cc [hunk 1]
```diff
@@ -12,20 +12,26 @@
...
 using namespace std;
 using namespace magneticfield;
 using namespace edm;
+using namespace angle_units::operators;
+
+using DDBox = dd4hep::Box;
+using DDTrap = dd4hep::Trap;
+using DDTubs = dd4hep::Tube;
+using DDCons = dd4hep::ConeSegment;
+using DDTruncTubs = dd4hep::TruncatedTube;
```
Comment (cvuosalo) [inline MagneticField/GeomBuilder/src/DD4hep_volumeHandle.cc:31]: If the above five lines were put into `DetectorDescription/DDCMS/interface/DDShapes.h`, then this PR wouldn't be needed. Or would there be some problem?

File: MagneticField/GeomBuilder/src/DD4hep_volumeHandle.cc [hunk 2]
```diff
@@ -197,9 +203,483 @@ namespace {
 
 using namespace cms::dd;
 
-#include "buildBox.icc"
-#include "buildTrap.icc"
-#include "buildTubs.icc"
-#include "buildCons.icc"
+/*
```
Comment (cvuosalo) [inline MagneticField/GeomBuilder/src/DD4hep_volumeHandle.cc:206]: I think the code duplication below should be avoided. I don't understand why the `icc` files can't be used as they are.

Comment (namapane) [reply MagneticField/GeomBuilder/src/DD4hep_volumeHandle.cc:206]: That was my question (see above) but Ianna pointed out to slight changes in interface, eg x() becoming halfX()
My suggestion is to extract the relevant parameters for each shape in *volumeHandle.cc  and pass them to the functions in the .icc files, as done already in buildPseudoTrap.icc. This way the actual implementation remains where it is and we just need to adjust the callers.

**Comment by slava77** [comment]
assign geometry

since quite a bit of the discussion and possible updates are related to geometry changes, I'd rather have a signature provided explicitly. The reco signature may follow after, assuming @namapane is happy with the changes.

**Comment by namapane** [comment]
@ianna - for what I'm concerned this PR should be recoded (ie closed and replaced by another one) refactoring the code that is being duplicated in this PR so that it does not need to be duplicated. It is a really easy task (just pass float parametes instead of shapes to the functions that are being replicated). As promised I plan to work on that. 
I do not see much point in releasing this PR as is by also moving here classes whose dependency was, in any case, introduced because of the migration to DD4Hep itself (and not by me). That it will make subsequent cleanup unnecessarily more complicated. 
That is just my opinion and you can of course go on if you prefer, Although I am sure that, on the path of release, others will object to the duplication of code snippets...

**Comment by ianna** [comment]
> @ianna - for what I'm concerned this PR should be recoded (ie closed and replaced by another one) refactoring the code that is being duplicated in this PR so that it does not need to be duplicated. It is a really easy task (just pass float parametes instead of shapes to the functions that are being replicated). As promised I plan to work on that.
> I do not see much point in releasing this PR as is by also moving here classes whose dependency was, in any case, introduced because of the migration to DD4Hep itself (and not by me). That it will make subsequent cleanup unnecessarily more complicated.
> That is just my opinion and you can of course go on if you prefer, Although I am sure that, on the path of release, others will object to the duplication of code snippets...

@namapane - thanks! OK, I'm closing this PR. Would you object moving the DDShapes.h/.cc under MagneticField/GeomBuilder/src/ to decouple the other pending PRs? Thanks.

**Comment by namapane** [comment]
I'm working on this right now

**Review by namapane — COMMENTED** [review]
Done, #31649.

**Pull request closed without merging**